bitcoin/src/blockdata/block.rs

```diff
@@ -275,38 +279,57 @@ impl Block {
         merkle_tree::calculate_root(hashes).map(|h| h.into())
     }
 
-    /// base_size == size of header + size of encoded transaction count.
-    fn base_size(&self) -> Weight {
-        Weight::from_wu_usize(80 + VarInt::from(self.txdata.len()).len())
+    /// Returns the weight of the block.
+    ///
+    /// > Block weight is defined as Base size * 3 + Total size.
+    ///
+    /// When calculating weight units we saturate at the bounds of `usize`, this should not happen
+    /// for any sane use case.
+    pub fn weight(&self) -> Weight {
+        // This is the exact definition of a weight unit, as defined by BIP-141 (quote above).
+        let wu = self.base_size().saturating_mul(3).saturating_add(self.total_size());
+        Weight::from_wu_usize(wu)
     }
 
-    /// Returns the size of the block.
+    /// Returns the base block size.
     ///
-    /// size == size of header + size of encoded transaction count + total size of transactions.
-    pub fn size(&self) -> usize {
-        let txs_size: usize = self.txdata.iter().map(Transaction::size).sum();
-        self.base_size().to_wu() as usize + txs_size
+    /// > Base size is the block size in bytes with the original transaction serialization without
+    /// > any witness-related data, as seen by a non-upgraded node.
+    ///
+    /// If overflow occurs we defensively return `usize::MAX`, this should not happen for any
+    /// sane use case.
+    fn base_size(&self) -> usize {
+        let mut size = Header::SIZE;
+
+        size = size.saturating_add(VarInt::from(self.txdata.len()).size());
+        for tx in self.txdata.iter() {
```

Comment:
nit: slightly prefer more rust idiomatic 

```rust
        self.txdata.iter().map(|tx| tx.base_size())
            .fold(size, |acc, elem| acc.saturating_add(elem))
```

Reply:
Some part of me does not like the inefficiency of having bounds check everytime in the loop when we are sure that we are not exceeding it. But I can live with it :)

Reply:
Same comment suggesting to convert loop to rust iterators where possible. More reading if you are interested: https://ipthomas.com/blog/2023/07/n-times-faster-than-c-where-n-128/

Reply:
Nice article.  Besides a performance increase, using bitwise operations also helps prevent against timing attacks by avoiding branching.  It would be possible to write a C version using bitwise operations as well I think.